### ISSUE TITLE
Improved Griffin-Lim via ADMM

### ISSUE THREAD

**Issue by PetrochukM** [librosa/librosa#1471]
**Is your feature request related to a problem? Please describe.**
Recently, there have been two papers that have demonstrated that Griffin-Lim ADMM has a higher quality than Fast Griffin Lim (the current Librosa implementation). They have also provided reference implementations, here:
https://github.com/phvial/PRBregDiv
https://github.com/dendisuhubdy/ADMMGLA

Here is the paper with the most robust study of various phase reconstruction algorithms across multiple corpuses and metrics:
https://arxiv.org/abs/2010.00392

**Describe the solution you'd like**
Could these solutions be ported over to Librosa, so we could have access to a higher quality Griffin-Lim algorithm? Griffin-Lim is used a lot in TTS, so it'd be used widely, I think!

**Comment by PetrochukM** [comment]
The algorithm is pretty simple too, feel free to take a look at the example code!

**Comment by bmcfee** [comment]
I haven't read the paper in depth, but it doesn't seem from the presented results that there's that much of a difference compared to accelerated GL.  Am I misreading something?

**Comment by PetrochukM** [comment]
Hey! 

It's certainly not dramatically better in their charts (I'm looking at GLADMM); however, it is overall better and has different performance characteristics depending on the dataset.
<img width="689" alt="image" src="https://user-images.githubusercontent.com/7424737/162094182-058641ab-d8a7-43f3-b4c6-b7904df843e9.png">
<img width="255" alt="image" src="https://user-images.githubusercontent.com/7424737/162094220-441b2db1-3f15-4ed8-b3b2-ebb93d73d0d1.png">

In our internal testing, we found that FGLA, especially at larger frame sizes on speech datasets, created a lot of distortion and was warbly. This didn't happen when we used standard GL or GLADMM. 

So, I think, it's worthwhile to have an alternative option, and the algorithm is conceptually simple.

**Comment by bmcfee** [comment]
Thanks for the details and context here.  I'm leaning against this for a couple of reasons, but could be persuaded.

1. The algorithm might be "conceptually simple", but in my experience, ADMM methods can be a bit brittle in practice when it comes to balancing the primal and dual objectives.  Getting an implementation that works well across a broad range of input data can be pretty subtle, e.g. with adaptive scaling of the augmenting parameter, etc.  This all can lead to a pretty complex implementation that will require significant expertise to understand and maintain.
2. We provide (F)GL mainly as a reference implementation.  There are plenty of other, better methods out there (PHGI comes to mind), but the goal for librosa isn't to implement every method.  Rather it's to provide usable and standardized implementations of basic methods that can be used either as baselines or as components in larger systems.  I'd prefer to not get into the game of aggregating all the latest and greatest methods into one package, as that again will become completely unmaintainable.

That said, the reference implementation for GLADMM seems to work, but I wouldn't necessarily call it "usable" as a software package.  (It's not an installable package from what I can tell, and doesn't have any unit tests or proper documentation.)    My suggestion here would be to improve that implementation as a stand-alone package that you could use with librosa, but not as part of the core library.

**Comment by PetrochukM** [comment]
Thanks for your input and for considering this!! 

I didn't know that ADMM is brittle in practice, thanks for the info! 

And that's fair. I think one thing that came up for me, was that I didn't know about [PHGI](https://arxiv.org/pdf/1609.00291.pdf) until you mentioned it just now. (Thank you!!) I have been doing TTS research for years, and I haven't seen any papers talking about it. Griffin-Lim is so important to TTS research, and yet, better methods like PHGI are not being used, maybe because there aren't any "usable" software packages with them? If it's not in `librosa`, it'd be great to see someone, somewhere, create a good software package that has a good state-of-the-art method that everyone could use as a go-to for phase reconstruction! I'd really appreciate that because Griffin-Lim is so central to TTS research, any improvements to it would be welcome!

**Comment by bmcfee** [comment]
> better methods like PHGI are not being used, maybe because there aren't any "usable" software packages with them? If it's not in `librosa`, it'd be great to see someone, somewhere, create a good software package that has a good state-of-the-art method that everyone could use as a go-to for phase reconstruction

Yes, though one of the issues with phgi is that it's not totally obvious (to me anyway) how to generalize it to support arbitrary window functions.  At least, that's why I haven't pursued an implementation.  I'd be open to a PR implementing it, but don't have the time or bandwidth to do it myself.

**Comment by phvial** [comment]
Hello! 

Thank you for the discussion above! I am a little late to this but here are my two cents:
We considered GLADMM as a baseline in our work not so much for its performances (even if they are interesting) but for its similarities with ADMM and (Fast) GL (and thus by extension with gradient methods). With regards to the reconstruction performances of GLADMM, I think it would be interesting to study more extensively its robustness to noise and the choice of STFT parameters in order to guarantee its (possible) superiority over FGL.

**Comment by bmcfee** [comment]
Coming back to this now, after having finally put some time into grokking the pghi method.  TLDR is that I think it should be reasonably easy to do, but that it's best as an initialization to an iterative procedure (eg GLA and friends) and not particularly great as a standalone method.  Summary below.

## Idea of the algorithm

(Sorry if I'm very late to the party on understanding how this all works!  I also haven't bothered looking at actual implementations of this method yet, so probably most of this is well known.  I'm documenting this mainly for myself while it's fresh in my head.)

The entire gist of the method is to 1) use differences of log spectral magnitude as an estimate of the phase gradient, 2) integrate these gradients to recover phase.

(1) works for gaussian windows, and should be an ok approximation for gaussian-looking windows (raised cosine etc).  Probably it's pretty bad for other window shapes, but also probably that's not very important in practice.

(2) is where the heap part comes in.  Rather than accumulating phase differentials like in a phase vocoder, the idea is to identify spectral peaks (where the gradient estimate should be reasonably stable), and do a breadth-first accumulation over the time-frequency grid.  (Pop the next largest magnitude point off, update its neighbors' phases by accumulation, and add them to the heap.)

This should be reasonably easy to implement using numba and a bounded heapq.  There are a few details to work out though.

## Details to work out

The algorithm is only locally stable, and there is nothing to ensure global consistency.  Each initial point in the heap is seeded with a phase of 0 (because why not), and if two exploration paths collide, there is absolutely no reason to expect them to produce a consistent estimate at the collision point.  This is likely the main reason why a postprocessing cleanup step is necessary (ie a few rounds of GL(A)).

A second reason for postprocessing cleanup is that the actual window function has an effect here that is only approximated by the gaussian assumption.  Probably the differences are minor in practice, but again, something that GL(A) postprocessing would clean up.

There are some magic numbers to sort out.  One could threshold the spectral magnitudes to prevent accumulation of gradients through low-energy regions where the estimates are unstable.  Relatedly, one could set a bound on how many steps out a gradient accumulation is allowed to propagate from its seed point.  (Or: one could do a phase reset around transients, but maybe this isn't so necessary with this approach as it is in a phase vocoder.  Something to experiment with anyway.)

## Proposed implementation

I think it would be reasonably straightforward to provide a pghi implementation with the following signature:

```python
def pghi(S, n_fft, hop_length, tolerance, heap_size, max_steps)
```

- S may be complex, in which case the angles encode initial estimates.  Otherwise angles of 0 are assumed for initialization.

- n_fft and hop_length are needed for the gradient estimations (group delay, instantaneous phase).  Note: this might need some more careful thinking if we want to support cqt/vqt as well as stft.

- tolerance would set the threshold below which gradient accumulation is blocked.

- heap_size could bound the number of frontier points kept around at any time.
 
- max_steps would limit how many moves out a gradient can accumulate.  Setting to None would allow unbounded propagation (provided tolerance criteria are met).



I don't think we need to support the full STFT parameter list like GL does.  Though I could imagine it being useful to know things like whether centering was used or the window length (in addition to frame length) so that we could, eg, avoid initializing the phase accumulation on frames in the padded regions where the assumptions of the gradient estimates will be very likely wrong.

Once that's all done, it should be easy to augment the GL API to support `init='pghi'` as well as the other existing options.
